crypto/provider_core.c

```diff
@@ -336,23 +375,69 @@ int ossl_provider_forall_loaded(OPENSSL_CTX *ctx,
 {
     int ret = 1;
     int i;
+    int use_fallbacks;
     struct provider_store_st *store = get_provider_store(ctx);
 
     if (store != NULL) {
         CRYPTO_THREAD_read_lock(store->lock);
-        for (i = 0; i < sk_OSSL_PROVIDER_num(store->providers); i++) {
-            OSSL_PROVIDER *prov = sk_OSSL_PROVIDER_value(store->providers, i);
+        for (use_fallbacks = 0; use_fallbacks < 2; use_fallbacks++) {
+            int num_provs = sk_OSSL_PROVIDER_num(store->providers);
+
+            /*
+             * if use_fallbacks is 1, it means there were no activated
+             * providers at all, and we need to auto-activate any fallback
+             * and try once more.
+             */
```

Comment:
I'm wondering if this is the right logic. It doesn't seem possible with how things are right now to be able to say "never use the default provider under any circumstances". I'm thinking that (perhaps) we should say that as soon as one provider has been activated we should never automatically activate the default provider even if the first provider is later de-activated, i.e. we only fallback if we have *never* had an active provider.

Reply:
Fair point.  That's easily solved with a flag in the store itself (i.e. in `struct provider_store_st`.